Lib/asyncore.py

```diff
@@ -143,21 +143,25 @@ def poll(timeout=0.0, map=None):
 
         r, w, e = select.select(r, w, e, timeout)
 
-        for fd in r:
-            obj = map.get(fd)
-            if obj is None:
+        # bpo-30931: Get objects from map before calling handlers,
+        # since handlers can modify map
+        robjs = [(fd, map[fd]) for fd in r]
+        wobjs = [(fd, map[fd]) for fd in w]
+        eobjs = [(fd, map[fd]) for fd in e]
+
+        for fd, obj in robjs:
+            if map.get(fd) is not obj:
+                # bpo-30931: obj has been closed by a previously executed
+                # handler.  Its file descriptor may have been reused by new
+                # dispatcher registered later.
```

Comment:
> Since the original check was if obj is None: shouldn't this be if map.get(fd) is not obj or obj is None:?

obj cannot be None.

The current code checks if fd is still in map because map can be modified while we iterate on ready file descriptors. My PR doesn't have this specific issue, since it iterates on a separated list.